src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +33,496 @@
+     *
+     * @return the next state to go to.
+     */
+    private AIState prepareDelivery()
+    {
+        final AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if (ownBuilding instanceof BuildingDeliveryman)
+        {
+            final AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                final boolean ableToDeliver = wareHouse.getTileEntity().checkInWareHouse(buildingToDeliver, false);
+                if (!ableToDeliver)
+                {
+                    buildingToDeliver.setOnGoingDelivery(false);
+                    return START_WORKING;
+                }
+                itemsToDeliver = new ArrayList<>(buildingToDeliver.getNeededItems());
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    private AIState gatherItemsFromWareHouse()
+    {
+        final AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if (ownBuilding instanceof BuildingDeliveryman)
+        {
+            final AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                if (itemsToDeliver.isEmpty() && hasTools(buildingToDeliver))
+                {
+                    return DELIVERY;
+                }
+
+                if(gatherItems(buildingToDeliver))
+                {
+                    setDelay(DUMP_AND_GATHER_DELAY);
+                    return GATHER_IN_WAREHOUSE;
+                }
+
+                ((BuildingDeliveryman) ownBuilding).setBuildingToDeliver(null);
+                itemsToDeliver.clear();
+                return START_WORKING;
+            }
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Check if the deliveryman has all the tools to make the delivery.
+     *
+     * @param buildingToDeliver the building to deliver to.
+     * @return true if is ready to deliver.
+     */
+    private boolean hasTools(@NotNull final AbstractBuilding buildingToDeliver)
+    {
+        final String requiredTool = buildingToDeliver.getRequiredTool();
+        if (requiredTool.isEmpty())
+        {
+            return true;
+        }
+
+        if (requiredTool.equals(Utils.PICKAXE))
+        {
+            return InventoryUtils.isPickaxeInTileEntity(worker.getInventoryCitizen(), buildingToDeliver.getNeededPickaxeLevel(), buildingToDeliver.getBuildingLevel());
+        }
+        return InventoryUtils.isToolInTileEntity(worker.getInventoryCitizen(), requiredTool, buildingToDeliver.getBuildingLevel());
+    }
+
+    /**
+     * Gather item from chest.
+     * Gathers only one stack of the item.
+     *
+     * @param buildingToDeliver building to deliver to.
+     * @return true if continue, false if not succuesful
+     */
+    private boolean gatherItems(@NotNull final AbstractBuilding buildingToDeliver)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 7, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "gatherItems" is 17 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)